### PR TITLE
Turn zope.meta into a real installable Python package

### LINKED ISSUE #97: RFE: packagize the metaconfiguration script
Every now and then I want to run the script to update a package, and then I realize that I can't just cd zope.sendmail && run something, I have to juggle two working directories somehow, and then I give up.

I'd like to be able to `pipx install -e ~/src/zopefoundation/meta` and have a runnable `zope-meta` (better name suggestions welcome) script in my PATH.

### PR DESCRIPTION
**Pull request by dataflake**
This PR turns zope.meta into a real Python package that can be installed like any other package. The scripts were changed slightly so they can be used as a console script entry point and the README was updated accordingly.

At the same time the packaging also allows running the zope.meta scripts against zope.meta itself, the results are also part of this PR.

My main goal is script re-use for custom applications. I would like to be able to create a simple script (or simple package) where I can import zope.meta code and override settings, like loading file templates from a different path or applying different settings from the zopefoundation defaults. This PR is just a first step towards that goal.

Fixes #97.

### CODE REVIEW

File: pyproject.toml
Comment (mgedmin) [inline pyproject.toml:3]: This part is surprising to me, especially given the `template = "pure-python"` in .meta.toml.

File: CHANGES.rst
Comment (mgedmin) [inline CHANGES.rst:5]: Maybe 

```suggestion
----------------
- Converted to a real installable package.
```

File: docs/hacking.rst
Comment (mgedmin) [inline docs/hacking.rst:11]: Do you plan to rename the repository after merging this?

I'm :+1: for that idea.

File: docs/hacking.rst
Comment (mgedmin) [inline docs/hacking.rst:72]: Do we have tests for this code now?  That is wonderful news!

File: docs/hacking.rst
Comment (mgedmin) [inline docs/hacking.rst:87]: This does not look truthful to me.  There's no message.py in src/zope/meta.

File: setup.py
Comment (mgedmin) [inline setup.py:91]:
```suggestion
            (
                'set-branch-protection-rules='
                'zope.meta.set_branch_protection_rules:main'
            ),
```

to make it more apparent that this is one list item split across two lines.

File: src/zope/meta/multi_call.py
Comment (mgedmin) [inline src/zope/meta/multi_call.py:40]: I think these two lines should be part of main() too.

**Review by mgedmin — APPROVED** [review]
Mostly LGTM!  Also, this PR makes me happy.

One script needs a bit more code to be moved inside main() so it won't do command-line argument parsing at run time.

The lack of actual tests makes me wonder about the utility of the tox.ini and docs/hacking.rst.  I think we should at least add an empty-ish tests.py so that anyone who wants to add a test will have a lower threshold of entry.  I don't think this PR needs to be blocked until we have actual tests -- as long as CI doesn't fail trying to enforce the 85% of coverage or whatever it has configured by default.

I would also like confirmation that all the references to github.com/zopefoundation/zope.meta are intentional and you plan to rename the repository.  I think GitHub will provide automatic redirection for people who use the old URL/have old git clones.

File: setup.py
Comment (mgedmin) [inline setup.py:62]: This is inconsistent with all the other places that already refer to https://github.com/zopefoundation/zope.meta

File: pyproject.toml
Comment (dataflake) [reply pyproject.toml]: The file was copied from another project, I fixed it.

File: CHANGES.rst
Comment (dataflake) [reply CHANGES.rst:5]: I added a change log entry

File: docs/hacking.rst
Comment (dataflake) [reply docs/hacking.rst]: Those pages were copied from another package. I removed all test output, that was not meant to suggest real test output, just demonstrate the command invocation.

File: docs/hacking.rst
Comment (dataflake) [reply docs/hacking.rst]: No, it is sample output, not reality, and it's gone now.

File: setup.py
Comment (dataflake) [reply setup.py]: Done

File: src/zope/meta/multi_call.py
Comment (dataflake) [reply src/zope/meta/multi_call.py]: I moved all the argparse bits into the respective `main` function

File: setup.py
Comment (dataflake) [reply setup.py]: That's all fixed.

**Review by d-maurer — APPROVED** [review]

File: docs/hacking.rst
Comment (dataflake) [reply docs/hacking.rst]: Fixed

File: setup.py
Comment (dataflake) [reply setup.py]: It still works, yes, but I just changed it to the zope.dev domain.

File: docs/hacking.rst
Comment (dataflake) [reply docs/hacking.rst]: The entire documentation text was copied from another project and then adapted where needed. I did not write this myself.

I have changed that passage to mention the contributor status and to discourage private forks, though.

**Review by mauritsvanrees — APPROVED** [review]
I did not try it, but the idea seems good, making it easier to apply zope/meta on a package without needing to have a git checkout of it.

This would even make it possible to add a scheduled GitHub workflow action that installs `zope.meta` and runs it on the repo and see if anything has changed.

File: setup.py
Comment (icemac) [inline setup.py:49]: This makes my heart bleed a bit: Previously the minimum requiered version for meta/config was Python 3.11. And I think that I was using Python 3.11+ syntax somewhere. Being forced to write Python 3.8 code in the once most advanced repo feels kind of sad.
But I see the point that it is easier for self updating the repository to support all the versions we currently support for ZF projects.

**Review by icemac — APPROVED** [review]
I really like the idea and admire the time you invested into this change. I already used `meta/config` for some OSS projects outside the ZF and making reuse easier will be a great win.

Though I have some hopefully helpful questions.

**Comment by icemac** [comment]
> This would even make it possible to add a scheduled GitHub workflow action that installs `zope.meta` and runs it on the repo and see if anything has changed.

You mean we could even make our repos self-updating. This is a really interesting idea. 😀

File: docs/_static/.gitignore
Comment (dataflake) [reply docs/_static/.gitignore:1]: It was so there's content in `docs/_static` and Sphinx would not warn about a missing folder. I removed it and commented out the `html_static_path` directive.

File: docs/hacking.rst
Comment (dataflake) [reply docs/hacking.rst]: I removed the list of Python versions.

File: setup.py
Comment (dataflake) [reply setup.py:49]: You were using `tomllib`, which (needlessly) introduced the hard Python 3.11 requirement. I replaced that with `tomlkit`, which is used everywhere else in the package. I find it more valuable to have a consistent set of supported versions everywhere. Less surprises.

Comment (dataflake) [reply setup.py:49]: And don't forget, we could not even run the configure script against this package without that consistent Python version support because there's no functionality for a different set.

**Comment by dataflake** [comment]
> I am not totally convinced about the renaming: In the archived packages we point to https://github.com/zopefoundation/meta to request unarchival. So maybe the code might go to zope.meta but the issues should stay in meta. Do you think this is reasonable and possible?

I am very surprised that you consider a potentially broken link in archived packages in any way important. I disagree and don't think that is important at all. Those are by definition outdated and abandoned.

I don't think it is reasonable and renaming a repository does not allow for keeping the issue tracker in the old place.

**Comment by dataflake** [comment]
> > This would even make it possible to add a scheduled GitHub workflow action that installs `zope.meta` and runs it on the repo and see if anything has changed.
> 
> You mean we could even make our repos self-updating. This is a really interesting idea. 😀

I am scared of this. Please don't. That would mean even the smallest and safest commit now risks a broken package and extra work because some of those pre-commit hooks go red.

**Comment by mauritsvanrees** [comment]
I suppose the repo could keep its `meta` name even when `setup.py` says `zope.meta`.

**Comment by dataflake** [comment]
> I suppose the repo could keep its `meta` name even when `setup.py` says `zope.meta`.

That would also mean we have another strange one-off, in this case with a package name/repository name mismatch. I'd prefer avoiding that.

**Comment by dataflake** [comment]
To finally get this off the ground I will retain the repository name and changed the documentation and configurations to make this work.

**Comment by mauritsvanrees** [comment]
Thanks a lot!

**Comment by icemac** [comment]
@dataflake Thank you for your work on this ticket, I really like the outcome!

**Pull request merged**
